clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsAssignmentInterface.java

```diff
@@ -281,17 +300,173 @@ public String toString() {
+
+            final Optional<KafkaException> error;
+
+            if (exceptionFromCallback.isPresent()) {
+                error = Optional.of(ConsumerUtils.maybeWrapAsKafkaException(exceptionFromCallback.get(), "Task assignment callback throws an error"));
+            } else {
+                error = Optional.empty();
+                reconciledAssignment.set(Assignment.EMPTY);
+            }
+
+            return new StreamsOnAllTasksLostCallbackCompletedEvent(future, error);
+        }
+    }
+
     public StreamsAssignmentInterface(UUID processId,
                                       Optional<HostInfo> endpoint,
                                       Map<String, Subtopology> subtopologyMap,
-                                      Map<String, String> clientTags
-    ) {
+                                      Map<String, String> clientTags) {
         this.processId = processId;
         this.endpoint = endpoint;
         this.subtopologyMap = subtopologyMap;
         this.taskLags = new HashMap<>();
         this.shutdownRequested = new AtomicBoolean(false);
         this.clientTags = clientTags;
+        this.streamsRebalanceEventProcessor = new StreamsRebalanceEventProcessor();
+    }
+
+    public void setOnTasksRevokedCallback(final Function<Set<StreamsAssignmentInterface.TaskId>, Optional<Exception>> onTasksRevokedCallback) {
+        this.onTasksRevokedCallback = Optional.ofNullable(onTasksRevokedCallback);
+    }
+
+    public void setOnTasksAssignedCallback(final Function<Assignment, Optional<Exception>> onTasksAssignedCallback) {
+        this.onTasksAssignedCallback = Optional.ofNullable(onTasksAssignedCallback);
+    }
+
+    public void setOnAllTasksLostCallback(final Supplier<Optional<Exception>> onAllTasksLostCallback) {
+        this.onAllTasksLostCallback = Optional.ofNullable(onAllTasksLostCallback);
+    }
+
+    public void setApplicationEventHandler(final ApplicationEventHandler applicationEventHandler) {
+        this.applicationEventHandler = applicationEventHandler;
+    }
+
+    public CompletableFuture<Void> requestOnTasksAssignedCallbackInvocation(final Assignment assignment) {
+        final StreamsOnTasksAssignedCallbackNeededEvent onTasksAssignedCallbackNeededEvent = new StreamsOnTasksAssignedCallbackNeededEvent(assignment);
+        onCallbackRequests.add(onTasksAssignedCallbackNeededEvent);
+        return onTasksAssignedCallbackNeededEvent.future();
+    }
+
+    public CompletableFuture<Void> requestOnTasksRevokedCallbackInvocation(final Set<StreamsAssignmentInterface.TaskId> activeTasksToRevoke) {
+        final StreamsOnTasksRevokedCallbackNeededEvent onTasksRevokedCallbackNeededEvent = new StreamsOnTasksRevokedCallbackNeededEvent(activeTasksToRevoke);
+        onCallbackRequests.add(onTasksRevokedCallbackNeededEvent);
+        return onTasksRevokedCallbackNeededEvent.future();
+    }
+
+    public CompletableFuture<Void> requestOnAllTasksLostCallbackInvocation() {
```

Comment:
Why all three `CompletableFuture` have no return value? I guess we're saying that `CompletableFuture.get()` doesn't throw an `ExcecutionException` everything succeeded?
Asking for my own understanding

Reply:
Basically, we use the futures only to bind code to them so that this code is executed when they complete. We pass the future from the background thread to the application thread which does something. Then when this something is done, we pass the future back to the background thread where it is completed and the code is executed. It is really just a way to pass around code and execute it at the right moment.